### PR TITLE
Alpha Sate

### PR DESCRIPTION
**Pull request by skipcheru**

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    app = Flask(__name__)
+    app.config.from_object(config[config_name])
+    config[config_name].init_app(app)
+
+    jwt = JWT(app, authenticate, identity)
+
+    db.init_app(app)
+    jwt.init_app(app)
+
+    from .auth.views import auth
```
Comment (kimobrian) [inline app/__init__.py]: Consider imports in the format `app.auth.views` , they are cleaner and only use imports within your code if you are trying to avoid cyclic imports otherwise put all imports at the top.

File: app/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+db = SQLAlchemy()
+
+
+def create_app(config_name):
+    app = Flask(__name__)
+    app.config.from_object(config[config_name])
+    config[config_name].init_app(app)
+
+    jwt = JWT(app, authenticate, identity)
```
Comment (kimobrian) [inline app/__init__.py]: Where are `identity` and `authenticate` declared?

File: app/auth/validate.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    def wrapper(f):
+        @wraps(f)
+        def decorated_function(*args, **kwargs):
+            if not request.is_json:
+                msg = "Request must be a valid json"
+                return jsonify({"error": msg}), 415
+            try:
+                data = request.get_json()
+                for argument in expected_args:
+                    if (argument not in data or not data.get(argument) or
```
Comment (kimobrian) [inline app/auth/validate.py]: These seem to do exactly the same thing, you can just use 1 of them: 
`argument not in data` and  `not data.get(argument)`

File: app/auth/validate.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                        msg = argument + ' required'
+                        return jsonify({'error': msg}), 400
+            except BadRequest:
+                msg = "Request must have valid data"
+                return jsonify({"error": msg}), 400
+            return f(*args, **kwargs)
+        return decorated_function
+    return wrapper
+
+from app.models import User
```
Comment (kimobrian) [inline app/auth/validate.py]: Use pep8 styling even for imports. check out [imports](https://www.python.org/dev/peps/pep-0008/#imports)

File: app/auth/views.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# from app.models import User
```
Comment (kimobrian) [inline app/auth/views.py]: It's good practice to remove commented out code.

File: app/auth/views.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    return jsonify({'message': 'Invalid credentials'}), 401
+
+
+# Register user
+@auth.route('/register', methods=['POST'])
+@validate_json('username', 'password')
+def register():
+    data = request.get_json()
+    username, password = data['username'], data['password']
+    user = User.query.filter_by(username=username).first()
```
Comment (kimobrian) [inline app/auth/views.py]: You can avoid the extra step of checking if a user exists just by setting `username` in your model and then instead of checking, just save and if your query violates the `uniqueness` rule, it will throw an `Integrity error` which you can catch and return a message within the `except` block.

File: app/auth/views.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    if user:
+        return jsonify({'error': 'User already exists'}), 409
+
+    user = User(username=username, password=password)
+    db.session.add(user)
+    db.session.commit()
+    return jsonify({'message': 'User registered successfully'}), 201
```
Comment (kimobrian) [inline app/auth/views.py]: How sure are youthat a user was saved successfully? Might consider having the `commit` within a `try...except` just in case of any errors during committing.

File: app/bucketlist/views.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+# Get all bucketlists and search bucketlists
+@bucketlists.route('/', methods=['GET'])
+@jwt_required()
+def get_bucketlist():
+    search = request.args.get('q', '')
+    page = request.args.get('page', 1)
+    limit = request.args.get('limit', 20)
+
+    results = BucketList.query.filter_by(user_id=current_identity.id).filter(
+        BucketList.name.ilike(f'%{search}%')).paginate(int(page), int(limit), False)
```
Comment (kimobrian) [inline app/bucketlist/views.py]: Use `try... except` block here to ensure that the `page` and `limit` are integers. If i send in `page` and `limit` as letters, your code will break at this point trying to cast strings.

File: app/bucketlist/views.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+
+# Add new bucketlist
+@bucketlists.route('/', methods=['POST'])
+@jwt_required()
+@validate_json('name')
+def new_bucketlist():
+    data = request.get_json()
+
+    bucket = BucketList.query.filter_by(name=data['name']).first()
```
Comment (kimobrian) [inline app/bucketlist/views.py]: What if the bucketlist belongs to another user. Give two user the liberty of having a bucketlist with the same name. Something like this will work for this: `bucket = BucketList.query.filter_by(name=data['name']).filter_by(user_id=current_identity.id).first()` which checks only bucketlists by current logged in user.

File: app/bucketlist/views.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    return jsonify({'message': 'BucketList added successfully'}), 201
+
+
+# Update single bucketlist
+@bucketlists.route('/<int:bucket_id>', methods=['PUT'])
+@jwt_required()
+@validate_json('name')
+def update_bucketlist(bucket_id):
+    data = request.get_json()
+    return any_request(request.method, BucketList, bucket_id, data)
```
Comment (kimobrian) [inline app/bucketlist/views.py]: Is `any_request` declared?

File: app/bucketlist/views.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+@jwt_required()
+def bucketlist(bucket_id):
+    return any_request(request.method, BucketList, bucket_id, None)
+
+
+# Get all bucketlist items
+@bucketlists.route('/<int:bucket_id>/items/', methods=['GET'])
+@jwt_required()
+def get_bucketlist_item(bucket_id):
+    all_items = [item.to_json() for item in
```
Comment (kimobrian) [inline app/bucketlist/views.py]: What if a user supplies a bucketlist id for someone else's bucketlist, they will be able to view them which is a violation of privacy. Consider something like `Item.query.filter_by(bucket_id=bucket_id).filter_by(user_id=current_identity.id)`

File: app/bucketlist/views.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    return jsonify(all_items), 200
+
+
+# Create new  bucketlist item
+@bucketlists.route('/<int:bucket_id>/items/', methods=['POST'])
+@jwt_required()
+@validate_json('name')
+def new_bucketlist_item(bucket_id):
+    data = request.get_json()
+    item = Item.query.filter_by(name=data['name']).first()
```
Comment (kimobrian) [inline app/bucketlist/views.py]: Allow different users to create bucketlist items with the same name and a single user to have bucketlist items with the same name but under different bucketlists.
`item = Item.query.filter_by(name=data['name']).filter_by(bucket_id=bucket_id).filter_by(bucket.user=current_user.id).first()`. Research on the `backref` option in your models to back reference the owner of a bucketlist.

File: tests/test_bucketlist.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
+from flask import json
+from base_class import TestBase
+from datetime import datetime
```
Comment (kimani-ndegwa) [inline tests/test_bucketlist.py]: Consider adding TESTS for cases where a user wants to create a duplicate bucketlist.
